app/services/logging/logger.js

```diff
@@ -1,17 +1,143 @@
+          format: this.getLogFormat(),
+        })
+      );
+    }
+
+    // Rolling and file are mutually exclusive, and rolling
+    // trumps the file transport.
+    if (
+      this.hasTransport(LoggerTransports.ROLLING) &&
+      this.loggerConfig.rotate &&
+      this.loggerConfig.path &&
+      this.loggerConfig.fileName
+    ) {
+      transports.push(
+        new this.winston.transports.DailyRotateFile({
+          // Check if the provided filename has the rolling
+          // date placeholder - if not, append it.
+          filename: /%DATE%/g.test(this.loggerConfig.fileName)
+            ? this.loggerConfig.fileName
+            : this.loggerConfig.fileName + "%DATE%",
+          dirname: this.loggerConfig.path,
+          datePattern: this.loggerConfig.rotate.datePattern,
+          zippedArchive: this.loggerConfig.rotate.zipped,
+          maxSize: this.loggerConfig.rotate.maxSize,
+          maxFiles: this.loggerConfig.rotate.maxFiles,
+          format: this.getLogFormat(),
+        })
+      );
+    } else if (
+      this.hasTransport(LoggerTransports.FILE) &&
+      this.loggerConfig.path &&
+      this.loggerConfig.fileName
+    ) {
+      transports.push(
+        new this.winston.transports.File({
+          // If a rolling date placeholder has somehow crept
+          // through, let's make sure to remove it here.
+          filename: path.join(
+            this.loggerConfig.path,
+            this.loggerConfig.fileName.replace(/%DATE%/g, "")
+          ),
+          format: this.getLogFormat(),
+        })
+      );
+    }
+
+    if (transports.length) {
+      this.logger = this.winston.createLogger({
+        level: this.loggerConfig.level,
+        transports,
+      });
+    }
+
+    return this;
+  }
+
+  exec(level, message, optionalParams) {
+    if (!level) {
+      level = "info";
+    }
+
+    // Fall back to default console implementation if
+    // no transports are specified.
+    if (!this.logger) {
+      if (optionalParams && optionalParams.length) {
+        console[level](message, optionalParams);
+      } else {
+        console[level](message);
+      }
+      return;
+    }
+
     if (optionalParams && optionalParams.length > 0) {
-      console.log(message, optionalParams);
+      this.logger.log({
+        level,
+        message,
+        meta: optionalParams,
+      });
     } else {
-      console.log(message);
+      this.logger.log({ level, message });
     }
   }
-}
+
+  log(message, ...optionalParams) {
+    this.exec("info", message, optionalParams);
+  }
+
+  debug(message, ...optionalParams) {
+    this.exec("debug", message, optionalParams);
+  }
+
+  info(message, ...optionalParams) {
+    this.exec("info", message, optionalParams);
+  }
+
+  warn(message, ...optionalParams) {
+    this.exec("warn", message, optionalParams);
+  }
+
+  error(message, ...optionalParams) {
+    this.exec("error", message, optionalParams);
+  }
+
+  getLogFormat() {
```

Comment:
I think this should also be a dependency - its own class that is injected into the logger and can be mocked / spied on as appropriate.

Reply:
Agreed. Going to implement this in a similar way to your `dbAdapterBase` and provide an abstract and a functional default.

Reply:
Well, the `Logger` class is the adapter in this instance, so everything is already in place. You *could* have an additional abstraction to remove the winston dependency from the logger class, but I think that's overkill here. We're much more likely to have multiple persistence stores than we are loggers (we're likely to just replace the logger than add another one).